### PR TITLE
Improve AES hardware-only check

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

Looking at #7384 (adds MBEDTLS_AES_USE_HARDWARE_ONLY), we complain with a #error message if this is set on x86 / x86_64 and MBEDTLS_AESNI_C is not set, and similar for Arm.

But if we set AES_USE_HARDWARE_ONLY on some other platform, there's no clean error about "not all prerequisites set", only an unexplained compile fail. This PR fixes that gap.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required - minor enhancement, no change to functionality
- [x] **backport**  not required - not in 2.28
- [x] **tests** not required - covered by existing

### CODE REVIEW

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM, I prefer the way this reverses things

File: library/aes.c [hunk 1]
```diff
@@ -34,23 +34,14 @@
...
+#if defined(MBEDTLS_AES_USE_HARDWARE_ONLY)
+#if !((defined(MBEDTLS_ARCH_IS_ARM64) && defined(MBEDTLS_AESCE_C)) || \
+      (defined(MBEDTLS_ARCH_IS_X64)   && defined(MBEDTLS_AESNI_C)) || \
+      (defined(MBEDTLS_ARCH_IS_X86)   && defined(MBEDTLS_AESNI_C)))
 #error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
 #endif
 #endif
 
-#if defined(MBEDTLS_ARCH_IS_X86)
-#if defined(MBEDTLS_AES_USE_HARDWARE_ONLY) && !defined(MBEDTLS_AESNI_C)
```
Comment (tom-cosgrove-arm) [inline library/aes.c:50]: Here we delete two `#if`s and only one `#else`, leading to

```
aes.c:2318:2: error: #endif without #if
#endif /* MBEDTLS_AES_C */
```

**Review by gilles-peskine-arm — DISMISSED** [review]
LGTM

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
And in case people hit the compiler warnings and search, the output you get when setting `MBEDTLS_AES_USE_HARDWARE_ONLY` on a platform that doesn't support it is

```
aes.c:41:2: error: "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
#error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
 ^
  CC    bignum_mod_raw.c
aes.c:619:15: warning: variable 'RK' set but not used [-Wunused-but-set-variable]
    uint32_t *RK;
              ^
aes.c:616:75: warning: unused parameter 'key' [-Wunused-parameter]
int mbedtls_aes_setkey_enc(mbedtls_aes_context *ctx, const unsigned char *key,
                                                                          ^
  CC    camellia.c
aes.c:1066:48: warning: unused parameter 'ctx' [-Wunused-parameter]
int mbedtls_aes_crypt_ecb(mbedtls_aes_context *ctx,
                                               ^
aes.c:1068:47: warning: unused parameter 'input' [-Wunused-parameter]
                          const unsigned char input[16],
                                              ^
aes.c:1069:41: warning: unused parameter 'output' [-Wunused-parameter]
                          unsigned char output[16])
                                        ^
```

**Comment by daverodgman** [comment]
> And in case people hit the compiler warnings and search, the output you get when setting `MBEDTLS_AES_USE_HARDWARE_ONLY` on a platform that doesn't support it is
> 
> ```
> aes.c:41:2: error: "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
> #error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
>  ^
>   CC    bignum_mod_raw.c
> aes.c:619:15: warning: variable 'RK' set but not used [-Wunused-but-set-variable]
>     uint32_t *RK;
>               ^
> aes.c:616:75: warning: unused parameter 'key' [-Wunused-parameter]
> int mbedtls_aes_setkey_enc(mbedtls_aes_context *ctx, const unsigned char *key,
>                                                                           ^
>   CC    camellia.c
> aes.c:1066:48: warning: unused parameter 'ctx' [-Wunused-parameter]
> int mbedtls_aes_crypt_ecb(mbedtls_aes_context *ctx,
>                                                ^
> aes.c:1068:47: warning: unused parameter 'input' [-Wunused-parameter]
>                           const unsigned char input[16],
>                                               ^
> aes.c:1069:41: warning: unused parameter 'output' [-Wunused-parameter]
>                           unsigned char output[16])
>                                         ^
> ```

It's annoying that the compiler generates all the other stuff instead of stopping at #error, but it's not really worth the code churn to fix it

**Comment by daverodgman** [comment]
Sorry, accidentally updated this by pushing to the wrong branch. I've reset to the point of review approval.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**